samples/WinForms/Dx12Panel.cs

```diff
@@ -0,0 +1,868 @@
+
+        // Root assets path
+        private readonly string _assetsPath;
+
+        private const uint FrameCount = 2;
+
+        // Pipeline objects
+        private D3D12_VIEWPORT _viewport;
+        private RECT _scissorRect;
+        private IDXGISwapChain3* _swapChain;
+        private ID3D12Device* _device;
+        private readonly ID3D12Resource*[] _renderTargets;
+        private ID3D12CommandAllocator* _commandAllocator;
+        private ID3D12CommandQueue* _commandQueue;
+        private ID3D12RootSignature* _rootSignature;
+        private ID3D12DescriptorHeap* _rtvHeap;
+        private ID3D12PipelineState* _pipelineState;
+        private ID3D12GraphicsCommandList* _commandList;
+        private uint _rtvDescriptorSize;
+        private bool _isOnColorToggle;
+
+        // App resources.
+        private ID3D12Resource* _vertexBuffer;
+        private D3D12_VERTEX_BUFFER_VIEW _vertexBufferView;
+
+        // Synchronization objects.
+        private uint _frameIndex;
+        private HANDLE _fenceEvent;
+        private ID3D12Fence* _fence;
+        private ulong _fenceValue;
+
+        public float AspectRatio => _aspectRatio;
+
+        public bool UseWarpDevice => _useWarpDevice;
+
+        private Vector2 _mousePriorPt;
+
+        #endregion data members
+
+        #region constructor
+
+        public DX12Panel(int width, int height, bool useWarpDevice = false) : base()
+        {
+            if (width <= 0 || height <= 0)
+            {
+                throw new ArgumentException("width and height must be positive and non-zero");
+            }
+
+            Width = width;
+            Height = height;
+            _assetsPath = GetAssetsPath();
+            _aspectRatio = width / ((float)height);
+
+            _renderTargets = new ID3D12Resource*[FrameCount];
+
+            _viewport = new D3D12_VIEWPORT {
+                TopLeftX = 0,
+                TopLeftY = 0,
+                Width = width,
+                Height = height,
+                MinDepth = D3D12_MIN_DEPTH,
+                MaxDepth = D3D12_MAX_DEPTH
+            };
+
+            _scissorRect = new RECT {
+                left = 0,
+                top = 0,
+                right = unchecked((int)width),
+                bottom = unchecked((int)height)
+            };
+
+            _useWarpDevice = useWarpDevice;
+            _isOnColorToggle = false;
+
+            Resize += OnResize;
+            MouseEnter += OnMouseEnter;
+            MouseLeave += OnMouseLeave;
+            MouseClick += OnMouseClick;
+            MouseDoubleClick += OnMouseDoubleClick;
+            MouseDown += OnMouseDown;
+            MouseUp += OnMouseUp;
+            MouseHover += OnMouseHover;
+            MouseMove += OnMouseMove;
+            MouseWheel += OnMouseWheel;
+            OnInit();
+        }
+
+        #endregion constructor
+
+        #region mouse event handling
+
+        private void OnMouseEnter(object? sender, EventArgs e)
```

Comment:
We should be overriding the existing virtual members we inherited from `Panel`.

Reply:
I'm also not sure these should be on here. There isn't really any specific functionality that we should be overriding and instead each sample should be hooking into the `DX12Panel` to change what is rendered, etc.

Reply:
I don't get what you are suggesting.  The use case I had in mind was that I drag rotate the triangle with the mouse. For that I need to implement OnMouseMove() and if I want the cursor to change to a cursor that indicates rotation then I need to set that in OnMouseEnter() and revert it in OnMouseLeave().  That is what these placeholders are there for.  How would you handle the cursor change and mouse drag rotation differently?

Reply:
`Panel` already defines `OnMouseMove` (and the others), so they should be overridden rather than redeclared. But, these also shouldn't be overridden until they are actually being used.

Reply:
OK. That is fine. Then I'll remove them for now.